openedx/core/lib/block_cache/block_structure_factory.py

```diff
@@ -0,0 +1,197 @@
+        block relations, transformer data, and block data.
+
+        Arguments:
+            block_structure (BlockStructure) - The block structure
+                that is to be serialized to the given cache.
+
+            cache (django.core.cache.backends.base.BaseCache) - The
+                cache into which cacheable data of the block structure
+                is to be serialized.
+        """
+        data_to_cache = (
+            block_structure._block_relations,
+            block_structure._transformer_data,
+            block_structure._block_data_map
+        )
+        zp_data_to_cache = zpickle(data_to_cache)
+        cache.set(
+            cls._encode_root_cache_key(block_structure.root_block_usage_key),
+            zp_data_to_cache
+        )
+        logger.debug(
+            "Wrote BlockStructure %s to cache, size: %s",
+            block_structure.root_block_usage_key,
+            len(zp_data_to_cache),
+        )
+
+    @classmethod
+    def create_from_cache(cls, root_block_usage_key, cache, transformers):
+        """
+        Deserializes and returns the block structure starting at
+        root_block_usage_key from the given cache, if it's found in the cache.
+
+        The given root_block_usage_key must equate the root_block_usage_key
+        previously passed to serialize_to_cache.
+
+        Arguments:
+            root_block_usage_key (UsageKey) - The usage_key for the root
+                of the block structure that is to be deserialized from
+                the given cache.
+
+            cache (django.core.cache.backends.base.BaseCache) - The
+                cache from which the block structure is to be
+                deserialized.
+
+            transformers ([BlockStructureTransformer]) - A list of
+                transformers for which the block structure will be
+                transformed.
+
+        Returns:
+            BlockStructure - The deserialized block structure starting
+            at root_block_usage_key, if found in the cache.
+
+            NoneType - If the root_block_usage_key is not found in the cache
+            or if the cached data is outdated for one or more of the
+            given transformers.
+        """
+
+        # Find root_block_usage_key in the cache.
+        zp_data_from_cache = cache.get(cls._encode_root_cache_key(root_block_usage_key))
+        if not zp_data_from_cache:
+            logger.debug(
+                "BlockStructure %r not found in the cache.",
+                root_block_usage_key,
+            )
+            return None
+        else:
+            logger.debug(
+                "Read BlockStructure %r from cache, size: %s",
+                root_block_usage_key,
+                len(zp_data_from_cache),
+            )
+
+        # Deserialize and construct the block structure.
+        block_relations, transformer_data, block_data_map = zunpickle(zp_data_from_cache)
+        block_structure = BlockStructureBlockData(root_block_usage_key)
```

Comment:
After further discussion, I've renamed `BlockStructureXBlockData` to `BlockStructureModulestoreData` so it's clearer in its name.